### PR TITLE
api/views: Create new viewership BigQuery API

### PR DESCRIPTION
**Pull request by victorges**
This implements the new Viewership V2 API in the analyzer service. It uses bigquery as the
data store and implements all the features proposed for the April project.

This implements DAT-62.

### CODE REVIEW

File: views/client.go [hunk 1]
```diff
@@ -6,28 +6,53 @@ import (
...
 	ID         string `json:"id"`
 	StartViews int64  `json:"startViews"`
 }
 
+type Metric struct {
+	PlaybackID string     `json:"playbackId"`
+	Time       *time.Time `json:"time,omitempty"`
+
+	Device      string `json:"device,omitempty"`
+	OS          string `json:"os,omitempty"`
```
Comment (ecmulli) [inline views/client.go:34]: add Browser

File: views/client.go [hunk 2]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+}
+
+func (c *Client) doViewsBigQuery(ctx context.Context, spec QuerySpec) ([]LDSMStateRow, error) {
+	query := squirrel.Select(
+		"countif(play_intent) as view_count",
+		"sum(playtime_ms) as playtime_ms",
+		"avg(ttff_ms) as ttff_ms",
+		"avg(rebuffer_ratio) as rebuffer_ratio",
+		"sum(error_count) as error_count",
+		"countif(exit_before_start) as exits_before_start",
```
Comment (ecmulli) [inline views/client.go:239]:
```suggestion
		"avg(if(exit_before_start, 1, 0)) as exits_before_start",
```
realized this is best understand as a percentage

File: views/client.go [hunk 3]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+	PlaytimeMs int64 `bigquery:"playtime_ms"`
+}
+
+func (c *Client) doViewsBigQuery(ctx context.Context, spec QuerySpec) ([]LDSMStateRow, error) {
+	query := squirrel.Select(
+		"countif(play_intent) as view_count",
+		"sum(playtime_ms) as playtime_ms",
+		"avg(ttff_ms) as ttff_ms",
+		"avg(rebuffer_ratio) as rebuffer_ratio",
+		"sum(error_count) as error_count",
```
Comment (ecmulli) [inline views/client.go:238]:
```suggestion
		"avg(error_count) as error_rate",
```
best expressed a rate but could do both - i'll leave that to you to decide

File: views/client.go [hunk 4]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+
+	// metric data
+	ViewCount  int64 `bigquery:"view_count"`
+	PlaytimeMs int64 `bigquery:"playtime_ms"`
+}
+
+func (c *Client) doViewsBigQuery(ctx context.Context, spec QuerySpec) ([]LDSMStateRow, error) {
+	query := squirrel.Select(
+		"countif(play_intent) as view_count",
+		"sum(playtime_ms) as playtime_ms",
```
Comment (ecmulli) [inline views/client.go:235]:
```suggestion
		"sum(playtime_ms) / 60000 as playtime_mins",
```
milliseconds is not interpretable. we should do minutes or hours. i think for the historical, minutes are best. the summary table uses hours but the bins can be small in this context and .25 hours is tough to interpret but 15 minutes is nice

File: views/client.go [hunk 5]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+
+var allowedTimeSteps = map[string]bool{
+	"hour":  true,
+	"day":   true,
+	"week":  true,
+	"month": true,
+	"year":  true,
+}
+
+type LDSMStateRow struct {
```
Comment (ecmulli) [inline views/client.go:214]: i'd renamed this. the table is called `viewership_events` something similar woudl make the most sense

File: views/client.go [hunk 6]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+	TimeInterval time.Time `bigquery:"time_interval"`
+	PlaybackID   string    `bigquery:"playback_id"`
+
+	// breakdown fields
+	Device      string `bigquery:"device"`
+	OS          string `bigquery:"os"`
```
Comment (ecmulli) [inline views/client.go:220]: add Browser `Browser string `bigquery:"browser"`

File: views/client.go [hunk 7]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+	UserAgent   string `bigquery:"user_agent_family"`
```
Comment (ecmulli) [inline views/client.go:221]: this one doesn't exist in this iteration. the ua-parser through js was a bit different from the python one i was using

File: views/client.go [hunk 8]
```diff
@@ -90,3 +191,141 @@ func startViewsQuery(playbackID, playbackRecordingID string) string {
...
+	Continent   string `bigquery:"playback_continent_name"`
+	Country     string `bigquery:"playback_country_name"`
+	Subdivision string `bigquery:"playback_subdivisions_name"`
+	TimeZone    string `bigquery:"playback_time_zone"`
```
Comment (ecmulli) [inline views/client.go:225]:
```suggestion
	TimeZone    string `bigquery:"playback_timezone"`
```

File: views/client.go (hunk 7 above)
Comment (ecmulli) [reply views/client.go:221]: there are the 4 fields that should be added 
-- device_type 
-- browser
-- browser_engine
-- cpu

File: views/client.go (hunk 3 above)
Comment (victorges) [reply views/client.go:247]: Let's keep it simple, doing only the rate as you said it's better. Should it be sth like this tho? Counting percentage of sessions with errors 🤔 
```suggestion
		"avg(if(error_count > 0, 1, 0)) as error_rate",
```

File: views/client.go (hunk 4 above)
Comment (victorges) [reply views/client.go:244]: Yeah I love that

File: views/client.go (hunk 3 above)
Comment (ecmulli) [reply views/client.go]: perfect, this looks good.

File: api/handler.go [hunk 9]
```diff
@@ -42,48 +67,104 @@ type apiHandler struct {
...
-func addViewershipHandlers(router *httprouter.Router, handler *apiHandler) {
-	opts := handler.opts
-	middlewares := []middleware{}
+func (h *apiHandler) viewershipHandler() chi.Router {
+	opts := h.opts
+
+	router := chi.NewRouter()
 	if opts.AuthURL != "" {
-		middlewares = append(middlewares, authorization(opts.AuthURL))
+		router.Use(authorization(opts.AuthURL))
```
Comment (gioelecerati) [inline api/handler.go:115]: Get total views shouldn't have the auth middleware

File: views/bigquery.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sql, args, nil
+}
+
+type ViewershipEventRow struct {
```
Comment (gioelecerati) [inline views/bigquery.go:155]: Would move this struct and the Summary one before the build functions or at the top

**Review by gioelecerati — APPROVED** [review]
LGTM with comments!

File: api/handler.go (hunk 9 above)
Comment (victorges) [reply api/handler.go:115]: Discussed offline and decided we should still keep it on the /total views API. Let's have the same authorization for this API as we have for the playback info endpoint.

**Pull request merged**